tests/test_pipeline.py

```diff
@@ -95,6 +96,54 @@ async def test_task_single(self):
         await task.run()
         assert task.has_finished()
 
+    async def test_task_observers(self):
+        frame_received = False
+
+        class CustomObserver(BaseObserver):
+            async def on_push_frame(self, data: FramePushed):
+                nonlocal frame_received
+
+                if isinstance(data.frame, TextFrame):
+                    frame_received = True
+
+        identity = IdentityFilter()
+        pipeline = Pipeline([identity])
+        task = PipelineTask(pipeline, observers=[CustomObserver()])
+        task.set_event_loop(asyncio.get_event_loop())
+
+        await task.queue_frames([TextFrame(text="Hello Downstream!"), EndFrame()])
+        await task.run()
+        assert frame_received
+
+    async def test_task_add_observer(self):
+        frame_received = False
+        frame_add_received = False
+
+        class CustomObserver(BaseObserver):
+            async def on_push_frame(self, data: FramePushed):
+                nonlocal frame_received
+
+                if isinstance(data.frame, TextFrame):
+                    frame_received = True
+
+        class CustomAddObserver(BaseObserver):
+            async def on_push_frame(self, data: FramePushed):
+                nonlocal frame_add_received
+
+                if isinstance(data.frame, TextFrame):
+                    frame_add_received = True
+
+        identity = IdentityFilter()
+        pipeline = Pipeline([identity])
+        task = PipelineTask(pipeline, observers=[CustomObserver()])
+        task.set_event_loop(asyncio.get_event_loop())
+        task.add_observer(CustomAddObserver())
+
+        await task.queue_frames([TextFrame(text="Hello Downstream!"), EndFrame()])
+        await task.run()
```

Comment:
Is it worth adding a unit test that adds an observer _after_ the pipeline is up and running? I can imagine needing to dynamically add an observer some time after the conversation starts.

Reply:
In fact...thinking about how I could've implemented the "deferred" `post_actions` execution in Pipecat Flows (where the `post_actions` only occur after the bot stops speaking, which in turn might be waiting on the user to speak first) using `add_observer()`:

* As soon as we set a node that requires deferred `post_actions`, we use `add_observer()` to detect when the bot stops speaking
* And then as soon as we handle the bot stopping speaking, we remove the observer...

...so should we also have a corresponding `remove_observer()`?

Reply:
> Is it worth adding a unit test that adds an observer _after_ the pipeline is up and running? I can imagine needing to dynamically add an observer some time after the conversation starts.

Yep, I can do that.

Reply:
I just added `remove_observer()` and also added a test for delayed `add_observer()` and also `remove_observer()`.